vm/src/function.rs

```diff
@@ -549,6 +549,36 @@ macro_rules! into_py_native_func_tuple {
                 })
             }
```

Comment:
Since we're essentially just using the generics on `IntoPyNativeFunc` as a tag to avoid conflicting generic impls, maybe we could unify the generic parameters into one FuncInfo parameter that's a tuple of `(Params, Return, VM)` and whatever else we might need later, to make it easier for functions like `new_rustfunc` that want to accept any sort of `IntoPyNativeFunc`. (Maybe even an [`HList`](https://docs.rs/frunk)? Probably not.)

Reply:
Ooh, or maybe we could have a wrapper trait that's basically
```rust
impl<P, R, VM, T> IntoPyNativeFunc for T
where T: IntoPyNativeFuncInner
{
    type Param = P;
    type Returns = R;
    type VMPresent = VM;
    fn into_func(self) -> PyNativeFunc {
        IntoPyNativeFuncInner::into_func(self)
    }
}
```

That way we can leave out the generics in most cases, and use `IntoPyNativeFunc<Params=...>` when we do care about some of them.

Reply:
To be honest, I didn't understand what you meant from this. will you please attach a commit into this PR?

Reply:
Nevermind, I thought that basically this would be possible:
```rust
pub trait IntoPyNativeFunc {
    type Params;
    type Result;
    fn into_func(self) -> PyNativeFunc;
}

impl<T, R, VM, F: IntoPyNativeFuncInner<T, R, VM>> IntoPyNativeFunc for F {
    type Params = T;
    type Result = R;
    fn into_func(self) -> PyNativeFunc {
        IntoPyNativeFuncInner::into_func(self)
    }
}

trait IntoPyNativeFuncInner<T, R, VM> {
    fn into_func(self) -> PyNativeFunc;
}
// implementing IntoPyNativeFuncInner for Fn(...) ...

// then that can be used in e.g. new_rustfunc without caring about the generics:
fn new_rustfunc(ctx, f: impl IntoPyNativeFunc) -> PyObjectRef { ... }
```
But that also has the issue of "potentially" multiple impls of `IntoPyNativeFuncInner` for the same type, even though we know that doesn't happen. We might be able to use the [`#[fundamental]` attribute](https://github.com/rust-lang/rfcs/blob/master/text/1023-rebalancing-coherence.md#errors-from-cargo-and-the-fundamental-attribute) once it's stabilized, but I guess it's not possible right now.